Fix bugs in codegen for aselect nodes where register not flagged as collected reference

rpshukla opened the pull request:
In the codegen for `aselect` nodes, if at least one of the possible values of the `aselect` is the address of a collected reference, then the result register for the `aselect` node should be flagged as a collected reference.

This PR addresses 2 existing issues related to collected references in `aselect` nodes:

  1. On some platforms, the evaluator for `aselect` nodes did not check for collected references or checked incorrectly. This PR ensures that the correct check is done and that `setContainsCollectedReference()` is called on the result register if necessary.

  2. On some platforms, `gprClobberEvaluate` was used to generate a register for the result of the `aselect` node. In some cases, `gprClobberEvaluate` will allocate a new register to hold the given value. In these cases, it should check to see if the value is a collected reference before moving it to the new register. If it is a collected reference, the new register should be flagged as containing a collected reference. This PR ensures this check is done on platforms that use `gprClobberEvaluate` while evaluating `aselect` nodes.

This also adds fatal asserts in evaluators for `aselect` nodes which will fail if the nodes have children that are internal pointers.

Issue: https://github.com/eclipse/openj9/issues/8397

Signed-off-by: Ryan Shukla <ryans@ibm.com>

rpshukla commented:
I have opened this as a draft PR since I am still working to verify that this fixes the original issue.

rpshukla commented:
fyi @andrewcraik

rpshukla commented:
Update: I have also added asserts in select evaluators which will fail if one of the children of the select node contains an internal pointer. Internal pointers cannot be handled since we cannot set the pinning array on the result register without knowing which side of the select will be taken.

andrewcraik commented on compiler/aarch64/codegen/ControlFlowEvaluator.cpp:
```diff
@@ -814,6 +814,14 @@ OMR::ARM64::TreeEvaluator::iselectEvaluator(TR::Node *node, TR::CodeGenerator *c
+   // on the result register without knowing which side of the select will be
+   // taken.
+   if (trueReg->containsInternalPointer() || falseReg->containsInternalPointer())
+      TR_ASSERT(false, "Select node should not have children containing internal pointers.\n");
```
Shouldn't this be assert fatal?

andrewcraik replied:
move the condition into the assert would also be cleaner

andrewcraik commented on compiler/riscv/codegen/ControlFlowEvaluator.cpp:
```diff
@@ -511,6 +511,13 @@ OMR::RV::TreeEvaluator::iselectEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   // on the result register without knowing which side of the select will be
+   // taken.
+   if (trueReg->containsInternalPointer() || falseReg->containsInternalPointer())
+      TR_ASSERT(false, "Select node should not have children containing internal pointers.\n");
```
assert fatal and why not just move the condition into the assert?

andrewcraik commented on compiler/p/codegen/ControlFlowEvaluator.cpp:
```diff
@@ -949,8 +949,16 @@ TR::Register *OMR::Power::TreeEvaluator::iselectEvaluator(TR::Node *node, TR::Co
+   // on the result register without knowing which side of the select will be
+   // taken.
+   if (trueReg->containsInternalPointer() || falseReg->containsInternalPointer())
+      TR_ASSERT(false, "Select node should not have children containing internal pointers.\n");
```
per above

andrewcraik commented:
Changes look reasonable, please un-WIP/draft when ready.

aviansie-ben commented on compiler/aarch64/codegen/ControlFlowEvaluator.cpp:
```diff
@@ -814,6 +814,14 @@ OMR::ARM64::TreeEvaluator::iselectEvaluator(TR::Node *node, TR::CodeGenerator *c
+   // on the result register without knowing which side of the select will be
+   // taken.
+   if (trueReg->containsInternalPointer() || falseReg->containsInternalPointer())
+      TR_ASSERT(false, "Select node should not have children containing internal pointers.\n");
```
In addition to what Andrew pointed out, I'd like to see this assert switched to use the new `TR_ASSERT_FATAL_WITH_NODE` macro so that we get node context printed on failure.

rpshukla commented:
Made the following updates:

- Switched to using new `TR_ASSERT_FATAL_WITH_NODE`
- Placed condition directly inside assert
- Dropped POWER changes as those have been added to PR #5497

andrewcraik commented on compiler/x/codegen/OMRCodeGenerator.cpp:
```diff
@@ -2032,10 +2032,14 @@ TR::Register *OMR::X86::CodeGenerator::gprClobberEvaluate(TR::Node * node, TR_X8
+         targetRegister->setContainsCollectedReference();
+      if (sourceRegister->containsInternalPointer())
+         {
+         targetRegister->setContainsInternalPointer();
```
This probably warrants a trace statement.

andrewcraik commented:
@fjeremic for x codegen change, @0xdaryl for the arm64 codegen change - who is looking after the risc-v codegen? would be good to get some eyes on that. I have left one comment on the x codegen related to tracing otherwise I think this looks reasonable.

0xdaryl commented:
The RISC-V changes are platform-neutral enough that I don't think they need a RISC-V expert to review.  However, I will just bring it to the attention of @janvrany in case he has comments.

rpshukla commented:
Added tracing everywhere `setContainsCollectedReference`, `setContainsInternalPointer`, and `setPinningArrayPointer` are called.

janvrany commented:
> who is looking after the risc-v codegen? 

That's me I guess. 

> ...in case he has comments.

LGTM.

andrewcraik commented:
@genie-omr build all

0xdaryl commented on compiler/aarch64/codegen/ControlFlowEvaluator.cpp:
```diff
@@ -814,6 +814,24 @@ OMR::ARM64::TreeEvaluator::iselectEvaluator(TR::Node *node, TR::CodeGenerator *c
+      !trueReg->containsInternalPointer() && !falseReg->containsInternalPointer(),
+      "Select nodes cannot have children that are internal pointers"
+   );
+   if (falseReg->containsCollectedReference())
```
I don't think this code under the `if` is actually useful on AArch64.  At this point in the function, `resultReg` should be the same as `trueReg`.  `trueReg` and `falseReg` are both the result of evaluation the true and false children, respectively.  So if `falseReg` is a collected reference then `trueReg` better be (and vice versa) otherwise the trees aren't properly formed.

andrewcraik replied:
The issue is that for NULL it can be collected or not collected so you have to do this copy in case you have a not collected null merging with a collected reference.

0xdaryl replied:
OK, for some reason I thought that an `aconst 0` evaluated into a register had the collected reference flag set.  It doesn't look like it though.  This change is fine then.
